File: server/excalibur_server/src/crypto/elliptic/noise_nk.py
from hashlib import sha256
from typing import Literal

from Crypto.Cipher import AES

from excalibur_server.src.crypto.elliptic.ristretto255 import Ristretto255
from excalibur_server.src.crypto.hkdf import HKDF

PROTOCOL_NAME = b"Noise_NK_Ristretto255_AESGCM_SHA256"


class NoiseNK:
    """
    [Noise-NK protocol](https://noiseprotocol.org/noise.html) implementation using Ristretto255,
    AES-GCM, and SHA256.
    """

    def __init__(self, server_pub: Ristretto255):
        """
        Initialize the Noise-NK protocol with the server's public key.

        :param server_pub: server's public key
        """

        self._server_pub = server_pub
        "Server public key"

        # Initialization
        self._h = sha256(PROTOCOL_NAME).digest()
        "Hash output"
        self._ck = self._h
        "Chaining key"

        # Pre-message
        self._h = sha256(self._h + self._server_pub.to_bytes()).digest()

    # Helper methods
    def _hkdf(self, ikm: bytes, num_outputs: Literal[2, 3]) -> tuple[bytes, ...]:
        hkdf = HKDF("sha256")
        temp_key = hkdf.hmac_hash(self._ck, ikm)

        output_1 = hkdf.hmac_hash(temp_key, b"\x01")
        output_2 = hkdf.hmac_hash(temp_key, output_1 + b"\x02")
        if num_outputs == 2:
            return output_1, output_2

        output_3 = hkdf.hmac_hash(temp_key, output_2 + b"\x03")
        return output_1, output_2, output_3

    def _mix_hash(self, data: bytes):
        self._h = sha256(self._h + data).digest()

    def _mix_key(self, ikm: bytes) -> bytes:
        self._ck, k = self._hkdf(ikm, 2)
        return k

    def _encrypt_and_hash(self, k: bytes, pt: bytes) -> bytes:
        cipher = AES.new(k, AES.MODE_GCM, nonce=b"\x00" * 12)
        ct, tag = cipher.encrypt_and_digest(pt)
        ct = ct + tag
        self._mix_hash(ct)
        return ct

    def _decrypt_and_hash(self, k: bytes, ct: bytes) -> bytes:
        self._mix_hash(ct)
        ciphertext, tag = ct[:-16], ct[-16:]
        cipher = AES.new(k, AES.MODE_GCM, nonce=b"\x00" * 12)
        return cipher.decrypt_and_verify(ciphertext, tag)

    # Main methods
    def message_c_to_s(self, client_keyshare_priv: int | None = None) -> tuple[Ristretto255, bytes]:
        """
        Client to server message.

        Message pattern is "e, es".

        :param client_keyshare_priv: optional client ephemeral private keyshare value
        :return: client's public keyshare and authentication tag
        """

        # Message pattern "e"
        client_keyshare_priv = client_keyshare_priv or Ristretto255.random_scalar()
        client_keyshare_pub = client_keyshare_priv * Ristretto255.GENERATOR
        self._mix_hash(client_keyshare_pub.to_bytes())

        # Message pattern "es"
        dh = (client_keyshare_priv * self._server_pub).to_bytes()
        k = self._mix_key(dh)

        # EncryptAndHash(empty)
        tag = self._encrypt_and_hash(k, b"")
        return client_keyshare_pub, tag

    def message_s_to_c(
        self,
        client_keyshare_pub: Ristretto255,
        client_tag: bytes,
        server_priv: int,
        server_keyshare_priv: int | None = None,
    ) -> tuple[Ristretto255, bytes, bytes]:
        """
        Server to client message.

        Message pattern is "e, ee".

        :param client_keyshare_pub: client's public keyshare
        :param client_tag: client's tag
        :param server_priv: server's private key
        :param server_keyshare_priv: optional server ephemeral private keyshare value
        :return: the server's public keyshare, the server's tag, and the shared secret
        """

        # Validate client message
        self._mix_hash(client_keyshare_pub.to_bytes())
        dh = (client_keyshare_pub * server_priv).to_bytes()
        k = self._mix_key(dh)
        self._decrypt_and_hash(k, client_tag)

        # Message pattern "e"
        server_keyshare_priv = server_keyshare_priv or Ristretto255.random_scalar()
        server_keyshare_pub = server_keyshare_priv * Ristretto255.GENERATOR
        self._mix_hash(server_keyshare_pub.to_bytes())

        # Message pattern "ee"
        dh = (client_keyshare_pub * server_keyshare_priv).to_bytes()
        k = self._mix_key(dh)

        # EncryptAndHash(empty)
        server_tag = self._encrypt_and_hash(k, b"")

        # Derive session key
        k_send, _ = self._hkdf(b"", 2)
        return server_keyshare_pub, server_tag, k_send

    def client_derive_session_key(
        self, client_keyshare_priv: int, server_keyshare_pub: Ristretto255, server_tag: bytes
    ) -> bytes:
        """
        Method to derive the session key on the client side.

        :param client_keyshare_priv: client's ephemeral private keyshare
        :param server_keyshare_pub: server's ephemeral public keyshare
        :param server_tag: server's tag
        :return: the shared secret
        """

        # Verify server message
        self._mix_hash(server_keyshare_pub.to_bytes())
        dh = (client_keyshare_priv * server_keyshare_pub).to_bytes()
        k = self._mix_key(dh)
        self._decrypt_and_hash(k, server_tag)

        # Derive session key
        k_send, _ = self._hkdf(b"", 2)
        return k_send

File: app/packages/main/src/lib/crypto/elliptic/noise-nk.ts
import createHash from "create-hash";

import HKDF from "@lib/crypto/hkdf";

import { GCMCipher, GCMDecipher } from "../cipher";
import Ristretto255 from "./ristretto255";

const PROTOCOL_NAME = Buffer.from("Noise_NK_Ristretto255_AESGCM_SHA256");

/**
 * [Noise-NK protocol](https://noiseprotocol.org/noise.html) implementation using Ristretto255,
 * AES-GCM, and SHA256.
 */
export default class NoiseNK {
    /** Server public key */
    private serverPub: Ristretto255;
    /** Hash output */
    private h: Buffer;
    /** Chaining key */
    private ck: Buffer;

    /**
     * Client's ephemeral private keyshare.
     *
     * This is set during the client-to-server message exchange.
     */
    private clientKeysharePriv: bigint = 0n;

    constructor(serverPub: Ristretto255) {
        this.serverPub = serverPub;

        // Initialization
        this.h = this._sha256(PROTOCOL_NAME).digest();
        this.ck = this.h;

        // Pre-message
        this.h = this._sha256(Buffer.concat([this.h, this.serverPub.toBytes()])).digest();
    }

    // Helper methods
    private _sha256(data: Uint8Array) {
        return createHash("sha256").update(data);
    }

    private _hkdf(ikm: Buffer, numOutputs: 2 | 3): Buffer[] {
        const hkdf = new HKDF("sha256");
        const tempKey = hkdf.hmacHash(this.ck, ikm);

        const output1 = hkdf.hmacHash(tempKey, Buffer.from([1]));
        const output2 = hkdf.hmacHash(tempKey, Buffer.concat([output1, Buffer.from([2])]));

        if (numOutputs === 2) {
            return [output1, output2];
        }

        const output3 = hkdf.hmacHash(tempKey, Buffer.concat([output2, Buffer.from([3])]));
        return [output1, output2, output3];
    }

    private _mixHash(data: Uint8Array) {
        this.h = this._sha256(Buffer.concat([this.h, data])).digest();
    }

    private _mixKey(ikm: Buffer): Buffer {
        const [newCk, k] = this._hkdf(ikm, 2);
        this.ck = newCk;
        return k;
    }

    private _encryptAndHash(k: Buffer, pt: Buffer): Buffer {
        const cipher = new GCMCipher("aes-256-gcm", k, Buffer.alloc(12));
        const ct = Buffer.concat([cipher.update(pt), cipher.final(), cipher.getAuthTag()]);
        this._mixHash(ct);
        return ct;
    }

    private _decryptAndHash(k: Buffer, ct: Buffer): Buffer {
        this._mixHash(ct);
        const ciphertext = ct.subarray(0, -16);
        const tag = ct.subarray(-16);

        const cipher = new GCMDecipher("aes-256-gcm", k, Buffer.alloc(12));
        cipher.setAuthTag(tag);
        const pt = Buffer.concat([cipher.update(ciphertext), cipher.final()]);
        return pt;
    }

    // Main methods
    /**
     * Client to server message.
     *
     * @param clientKeysharePriv optional client ephemeral private keyshare value
     * @returns client's public keyshare and authentication tag
     */
    messageCToS(clientKeysharePriv?: bigint): [Ristretto255, Buffer] {
        // Message pattern "e"
        this.clientKeysharePriv = clientKeysharePriv ?? Ristretto255.randomScalar();
        const clientKeysharePub = Ristretto255.GENERATOR.mul(this.clientKeysharePriv);
        this._mixHash(clientKeysharePub.toBytes());

        // Message pattern "es"
        const dh = this.serverPub.mul(this.clientKeysharePriv).toBytes();
        const k = this._mixKey(Buffer.from(dh));

        // EncryptAndHash(empty)
        const tag = this._encryptAndHash(k, Buffer.alloc(0));
        return [clientKeysharePub, tag];
    }

    /**
     * Derive the session key on the client side.
     *
     * @param serverKeysharePub server's ephemeral public keyshare
     * @param serverTag server's tag
     * @returns the shared secret
     */
    deriveSessionKey(serverKeysharePub: Ristretto255, serverTag: Buffer): Buffer {
        // Verify server message
        this._mixHash(serverKeysharePub.toBytes());
        const dh = serverKeysharePub.mul(this.clientKeysharePriv).toBytes();
        const k = this._mixKey(Buffer.from(dh));
        this._decryptAndHash(k, serverTag);

        // Derive session key
        const [kSend, _] = this._hkdf(Buffer.alloc(0), 2);
        return kSend;
    }
}
